git: Check push config before falling back to branch remote

errmayank opened the pull request:
Closes #26649

Release Notes:

- Git will now check pushRemote and pushDefault configurations before falling back to branch remote

cole-miller commented:
Thanks! According to the [docs](https://git-scm.com/docs/git-rev-parse) it seems like we should be able to use `git rev-parse @{push}` to get the information about where `git push` should be directed without running `git config` multiple times--would you be interested in revising this to use that approach?

cole-miller commented:
Thanks for the information! I think it makes sense that we'd honor `push.default` here, so we act the same way as if you did `git push` at the command line (except that we show a picker instead of just erroring if you're in `simple` mode and a remote couldn't be determined). Are there any other issues with using `@{push}`?

The other thing is that I think we should separate out a `GitRepository` method for this, since we shouldn't do `git rev-parse @{push}` for all calls to `GitRepository::remotes` (e.g. if you do `git: fetch from` we want to show a list of all remotes regardless of the current branch/push config). Could you implement that?

errmayank commented:
> Are there any other issues with using `@{push}`?

Nope, I think that's all.

> The other thing is that I think we should separate out a `GitRepository` method for this, since we shouldn't do `git rev-parse @{push}` for all calls to `GitRepository::remotes` (e.g. if you do `git: fetch from` we want to show a list of all remotes regardless of the current branch/push config). Could you implement that?

Makes sense, I've updated it but instead of creating a separate function I added a `RemoteOperationKind` enum which I think makes it a bit cleaner, let me know what you think.

errmayank committed:
Create separate functions to get push, branch and all remotes

kubkon reviewed (changes requested):
Looks great, thanks! In addition to some nits I left in my review, we still need to address missing support for remote projects. The entry point to that would be `handle_get_remotes` in `git_store.rs` which should unpack `is_push` from the message envelope. This will require adding `is_push` as a `bool` field in `git.proto` for `message GetRemotes`.

errmayank committed:
Update crates/project/src/git_store.rs

Co-authored-by: Jakub Konka <kubkon@jakubkonka.com>

errmayank committed:
Update crates/git/src/repository.rs

Co-authored-by: Jakub Konka <kubkon@jakubkonka.com>

errmayank committed:
Update crates/git/src/repository.rs

Co-authored-by: Jakub Konka <kubkon@jakubkonka.com>

errmayank committed:
Update handle_get_remotes to use is_push

kubkon reviewed (approved):
Excellent work, thanks!
